### ISSUE TITLE
Seneca 3.0 Logging mutates parameters and makes it hard to use custom loggers.

### ISSUE THREAD

**Issue by Yumonra07** [senecajs/seneca#531]
## Mutation of log parameters
Current logging implementation gives Seneca log metadata(like `when`, `level`, `seneca`) by mutating the logged parameter. 

```
const param1 = { a: 'test object' };
seneca.log.info(param1);
console.log("Mutated ==>", param1);

// Seneca logs...
// Mutated ==> { a: 'test object', level: 'info', seneca: 'mazjqcko51bp/1473403037008/6643/3.0.0/-', when: 1473403037085 }
```

This isn't done by the default logger but by rather a middleware that i couldn't find a way to replace. So i can't fix this even if i were to use a custom logger.

```
function Logger(){}
Logger.preload = () => { return { extend: { logger: () => {}}}; };
seneca.use(Logger);
```

This middleware also doesn't care about the type of the object. Be it an array, exception or a simple object, it mutates it without any additional logic. 

```
const test = ['a', 'b', 'c'];
seneca.log.info(test);
console.log(test);
// [ 'a', 'b', 'c', level: 'info', seneca: 'b69b9dv72hhe/1473403656942/6907/3.0.0/-', when: 1473403657020 ]
```

This is problematic because most Javascript libraries doesn't expect this kind of objects. I wouldn't expect an Array with string key values if i were to write a library. Most of the other devs doesn't neither since using `JSON.stringify` to stringify the above array will result in `["a", "b", "c"]`, making the metadata disappear. This is why i had to sanitize log params before logging them in my custom logger.

### Solution Suggestion
I think the metadata should be an additional param to the custom logger next to `context`. This way we can choose to use the metadata or throw it away. This will also make the problem of objects having same attribute as Seneca log metadata. We can either choose to override the object attributes with Seneca metadata or do something else but either way this logic would be moved to the logger implementation. 

## Log level filtering
An additional problem i'm having when implementing a custom logger is that; there isn't an easy way to see if my custom logger should handle the given log. I think filtering logs by their log levels should be a part of the Seneca core, rather than being something we should separately do in our custom logger. Basic logger does this by using `patrun` to check if it should log the given payload. 

### Solution Suggestion
I think an utility function that creates a patrun by a given Seneca log level option should be exposed.

**Comment by mcdonnelldean** [comment]
@Yumonra07 Thank you for the very detailed feedback. @rjrodger you should have a read through here, it relates to our conversation around logging on slack yesterday.

@Yumonra07 Similar feedback has been made yesterday too and I really like your idea of meta / non user object extension. I have a question though. Given a 'meta' object passed in would you expect keys like level to be added to meta too? So that the only thing in the payload is user entered content?

The string array dump is a bug I will try to fix before the weekend is up. It was raised yesterday too. It should always log an object, `{data: ['x', 'y']}` in the case of `seneca.log.info('x','y')`.

Agree on log level filtering. We have someone working on a common module right now. The idea is we will have a baseline filter module that you can use or wrap in your own module. It will be drop in usable and be considered the 'core' set of log filters you should support. It won't be core in a sense that it will live in an external module, but it will be added to both the core and legacy logger plus any other loggers we have made as a dependency and improve common parity.

**Comment by dgonzalez** [comment]
Hi @Yumonra07. I am working in the filtering. This is the repo where it will be: https://github.com/senecajs/seneca-log-filter.

Please bear in mind that it is work in progress but if you have any suggestion, please shout.

Thanks!

**Comment by rjrodger** [comment]
@Yumonra07 thanks!

Core logging just outputs the full internal message, including meta data. Loggers can then decide what to do with it.

The seneca-log-filter will need to include an option to remove the meta data for this use case.
https://github.com/senecajs/seneca/issues/531

**Comment by Yumonra07** [comment]
@mcdonnelldean Yes i would expect to get log level in metadata aswell. Then i could implement a different strategy for each level of logs. Making the payload equal to the user entered content is what makes sense for me. Because then i wouldn't worry about attributes in my log objects which could clash with Seneca metadata.

Lets say i want to log the time in ISO8601 format on a field called `time`; there are 2 problems now. First: i wouldn't want to parse the `when` field unless i have to log that object(seneca-log-filter is awesome for this!). Second: i would have to make sure that i'm not logging any object that contains `level` or `when` fields.

Seperating the user entered content/metadata and `seneca-log-filter`, solves all of my problems. By doing this, i could implement different behaviour for each log level, Seneca plugin etc. by taking a look at the meta data and i wouldn't have to worry about mutation.
